### PR TITLE
[MHLO] add DynamicBroadcastInDimOp canonicalize pattern

### PR DESCRIPTION
**Pull request by qingyunqu**
* replace constant shape's `dynamic_broadcast_in_dim` with `broadcast_in_dim`

### CODE REVIEW

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc [hunk 1]
```diff
@@ -1724,13 +1724,39 @@ class DynamicBroadcastInDimOpNotActuallyDynamic
...
+    if (type.hasStaticShape()) {
+      rewriter.replaceOpWithNewOp<BroadcastInDimOp>(op, type, op.operand(),
+                                                    op.broadcast_dimensions());
+      return success();
+    } else if (outputDimOp &&
+               outputDimOp->hasTrait<mlir::OpTrait::ConstantLike>()) {
+      SmallVector<int64_t, 4> outputShape;
+      DenseIntElementsAttr shapeAttr;
+      if (auto constOp = llvm::dyn_cast<mhlo::ConstOp>(outputDimOp)) {
+        shapeAttr = constOp.value().cast<DenseIntElementsAttr>();
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:1743]: Can you write it this way: `  if (matchPattern(outputDimOp, m_Constant(&shapeAttr)) ` ; it avoids encoding the exact constant operation.

It should catch both cases.

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc [hunk 2]
```diff
@@ -1724,13 +1724,39 @@ class DynamicBroadcastInDimOpNotActuallyDynamic
...
+      } else if (auto shapeOp =
+                     llvm::dyn_cast<shape::ConstShapeOp>(outputDimOp)) {
+        shapeAttr = shapeOp.getShapeAttr();
+      } else {
+        return rewriter.notifyMatchFailure(op, "unsupport constant");
+      }
+      for (APInt shape : shapeAttr) {
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:1750]: I would think we could write `for (int64_t shape : shapeAttr.getValues<int64_t>())` ?

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc [hunk 3]
```diff
@@ -1724,13 +1724,39 @@ class DynamicBroadcastInDimOpNotActuallyDynamic
...
+        op.output_dimensions().getType().dyn_cast<RankedTensorType>();
+    if (!type || !operandType || !outputDimType ||
         !operandType.hasStaticShape()) {
-      return rewriter.notifyMatchFailure(op, "requires static shape");
+      return rewriter.notifyMatchFailure(op, "requires operand static shape");
     }
-    rewriter.replaceOpWithNewOp<BroadcastInDimOp>(
-        op, op.getType(), op.operand(), op.broadcast_dimensions());
-    return success();
+    if (type.hasStaticShape()) {
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:1734]: Can you add a comment documenting the pattern you're catching and how they are replaced?

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc (hunk 2 above)
Comment (qingyunqu) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:1750]: Maybe we couldn't write like this. There is `%0 = mhlo.constant dense<[4, 32]> : tensor<2xi32>`. 
`getValues<int64_t>` will assert failing in this pattern.

Comment (joker-eph) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc]: Well `getValues<int32_t>()` then :)
(I didn't check the actual element type used before)

Comment (qingyunqu) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc]: Also, there has pattern: `%0 = shape.const_shape [4, 32] : tensor<2xindex>`.  
So I think using `APInt` is suitable.

**Review by sherhut — APPROVED** [review]

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc [hunk 4]
```diff
@@ -1724,13 +1724,35 @@ class DynamicBroadcastInDimOpNotActuallyDynamic
                                 PatternRewriter& rewriter) const override {
     auto type = op.getType().dyn_cast<RankedTensorType>();
     auto operandType = op.operand().getType().dyn_cast<RankedTensorType>();
-    if (!type || !type.hasStaticShape() || !operandType ||
-        !operandType.hasStaticShape()) {
-      return rewriter.notifyMatchFailure(op, "requires static shape");
+    auto outputDimOp = op.output_dimensions().getDefiningOp();
+    auto outputDimType =
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:1728]:
```
tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc#2915:10: error: unused variable 'outputDimType' [-Werror,-Wunused-variable]
    auto outputDimType =
```

**Review by joker-eph — COMMENTED** [review]
Can you rebase and fix the unused variable.

**Comment by qingyunqu** [comment]
> Can you rebase and fix the unused variable.

ok

**Comment by qingyunqu** [comment]
I think there are some test error in master branch? Should I fix this? @joker-eph 
image

**Comment by joker-eph** [comment]
Are these tests failing because of your change? Otherwise you may just have been unlucky and rebasing would likely clear this.

**Comment by qingyunqu** [comment]
> Are these tests failing because of your change? Otherwise you may just have been unlucky and rebasing would likely clear this.

These tests failed also in **master branch**.

**Pull request closed without merging**